### PR TITLE
Comments

### PR DESCRIPTION
**Pull request by efatsi**
Pull Request # 2. Comments are fully functional and recursive. The app deals with users much better now, redirects a user to where they were when they login. This pull request also has all of the fixes proposed in the first pull request, as I ignorantly made those changes in later commits instead of to the code in the pull request.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -4,17 +4,31 @@ class ApplicationController < ActionController::Base
...
 
   def logged_in?
   	current_user
   end
-  helper_method :logged_in?
-
-  def check_login
-  	redirect_to login_url, alert: "You need to log in to view this page." if current_user.nil?
+  
+  def require_user
```
Comment (brianjlandau) [inline app/controllers/application_controller.rb]: It might be worth putting all this auth code in it's own module that gets included into `ApplicationController`. It helps with code clarity and maintainability as you move forward. It also makes it easier to switch auth providers in the future.

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -1,84 +1,54 @@
...
-      format.json { render json: @comments }
-    end
-  end
-
-  # GET /comments/1
-  # GET /comments/1.json
-  def show
+  
+  before_filter :assign_comment, :only => [:show, :destroy]
+  def assign_comment
```
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: You should move this down and make it a private method. Public methods are exposed potentially as actions.

File: app/controllers/articles_controller.rb [hunk 3]
```diff
@@ -83,4 +69,9 @@ def destroy
       format.json { head :no_content }
     end
   end
+  
+  def assign_article
```
Comment (brianjlandau) [inline app/controllers/articles_controller.rb]: This should be a private method.

File: app/controllers/comments_controller.rb [hunk 4]
```diff
@@ -1,84 +1,54 @@
...
-    respond_to do |format|
-      format.html { redirect_to comments_url }
-      format.json { head :no_content }
-    end
+    # flash[:notice] = "Successfully destroyed comment."
+    redirect_to comments_url
   end
+  
+  private
+  def find_commentable
```
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: :+1: for this abstraction

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -1,84 +1,54 @@
...
+  # open this up to users who made them  
   def destroy
-    @comment = Comment.find(params[:id])
     @comment.destroy
-
-    respond_to do |format|
-      format.html { redirect_to comments_url }
-      format.json { head :no_content }
-    end
+    # flash[:notice] = "Successfully destroyed comment."
```
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: Version control is your friend here. There should be a really compelling reason to leave something as a commented line instead of just removing it. Generally, you just remove it and if you need it back later you just get it back via your repo's history.

File: app/controllers/comments_controller.rb [hunk 6]
```diff
@@ -1,84 +1,54 @@
...
-        format.html { redirect_to @comment, notice: 'Comment was successfully created.' }
-        format.json { render json: @comment, status: :created, location: @comment }
-      else
-        format.html { render action: "new" }
-        format.json { render json: @comment.errors, status: :unprocessable_entity }
-      end
+    @commentable = find_commentable
+    @comment = @commentable.comments.build(params[:comment].merge(:user_id => current_user.id))
+    if @comment.save
+      # flash[:notice] = "Successfully created comment."
```
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: Same thing as bellow about commented out lines.

File: app/controllers/comments_controller.rb [hunk 7]
```diff
@@ -1,84 +1,54 @@
...
-
-    respond_to do |format|
-      format.html # new.html.erb
-      format.json { render json: @comment }
+  
+  load_and_authorize_resource
+  
+  def index
+    @commentable = find_commentable
+    if !@commentable.nil?
```
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: Instead of an an `if !` you should use an `unless` or you should keep the `if` and not have the `!` and flip the blocks inside.

File: app/controllers/comments_controller.rb [hunk 8]
```diff
@@ -1,84 +1,54 @@
...
-    respond_to do |format|
-      if @comment.save
-        format.html { redirect_to @comment, notice: 'Comment was successfully created.' }
-        format.json { render json: @comment, status: :created, location: @comment }
-      else
-        format.html { render action: "new" }
-        format.json { render json: @comment.errors, status: :unprocessable_entity }
-      end
+    @commentable = find_commentable
+    @comment = @commentable.comments.build(params[:comment].merge(:user_id => current_user.id))
```
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: :+1: good job applying what we talked about the other day.

File: app/controllers/comments_controller.rb [hunk 9]
```diff
@@ -1,84 +1,54 @@
...
-    @comment = Comment.new
-
-    respond_to do |format|
-      format.html # new.html.erb
-      format.json { render json: @comment }
+  
+  load_and_authorize_resource
+  
+  def index
+    @commentable = find_commentable
```
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: You do this a couple times. It should probably be a before filter.

File: app/controllers/comments_controller.rb [hunk 10]
```diff
@@ -1,84 +1,54 @@
...
-        format.html { render action: "new" }
-        format.json { render json: @comment.errors, status: :unprocessable_entity }
-      end
+    @commentable = find_commentable
+    @comment = @commentable.comments.build(params[:comment].merge(:user_id => current_user.id))
+    if @comment.save
+      # flash[:notice] = "Successfully created comment."
+      commenting_on = @comment.commentable_type.pluralize
+      id = @comment.commentable_id
+      eval "redirect_to #{@comment.commentable_type.downcase}_path(:id=>#{id})"
```
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: We should avoid `eval` where ever we can. In this case we could do something like this:

``` ruby
redirect_to self.send("#{@comment.commentable_type.downcase}_path", @comment.commentable_id)
```

File: app/models/article.rb [hunk 11]
```diff
@@ -1,25 +1,51 @@
...
   attr_accessible :link, :points, :title, :user_id
   
   # Relationships
   belongs_to :user
+  has_many :comments, :as => :commentable
   
   # Validations
-  validates_presence_of :title, :link, :user_id, :date
+  validates_presence_of :title, :link, :user_id
+  validates_format_of :link, :with => /^(http|https):\/\/([a-z0-9]*[\-\.])?([a-z0-9]+*\.[a-z]{2,5})(:[0-9]{1,5})?(\/.*)?$/
```
Comment (brianjlandau) [inline app/models/article.rb]: This is probably easier:

``` ruby
validates_format_of :link, :with => URI.regexp(['http', 'https'])
```

File: app/models/article.rb [hunk 12]
```diff
@@ -1,25 +1,51 @@
...
   
-  def show_user
-    u = self.get_user
-    u.username
-  end
-  
-  def get_user
-    User.find(self.user_id)
-  end
+  # This did not work :(
```
Comment (brianjlandau) [inline app/models/article.rb]: Delete

File: app/models/article.rb [hunk 13]
```diff
@@ -1,25 +1,51 @@
 class Article < ActiveRecord::Base
+  
+  delegate :username, :to => :user
+  
+  # before_validation :fill_user
```
Comment (brianjlandau) [inline app/models/article.rb]: Delete

File: app/models/article.rb [hunk 14]
```diff
@@ -1,25 +1,51 @@
...
+	def comment_count
+	  count = comments.length
+	  comments.each do |c|
+	    count += c.comment_count
+	  end
+	  count
+  end
+  
+  def adjust_link
+    if self.link.match(/^(http|https):\/\/([a-z0-9]*[\-\.])?([a-z0-9]+*\.[a-z]{2,5})(:[0-9]{1,5})?(\/.*)?$/).nil? and !self.link.nil?
```
Comment (brianjlandau) [inline app/models/article.rb]: You can probably use `URI.regexp(['http', 'https'])` here too.

File: app/models/article.rb (hunk 11 above)
Comment (brianjlandau) [inline app/models/article.rb]: Although that doesn't require that, that the URL is all that's in the string. We can talk about how to achieve that if you want.

File: app/models/article.rb [hunk 15]
```diff
@@ -1,25 +1,51 @@
...
+      self.link = "http://" + self.link
+    end    
+  end
+  
+	def short_link
```
Comment (brianjlandau) [inline app/models/article.rb]: Would this work?

``` ruby
def short_link
  link.gsub(/\Ahttps?/i, '')
end
```

File: app/models/comment.rb [hunk 16]
```diff
@@ -1,3 +1,32 @@
...
+  belongs_to :user
+  has_many :comments, :as => :commentable
+  
+  # Validations
+  validates_presence_of :content, :user
+  
+  # Methods
+  
+  # This works but I don't think is very clean, outputs a string like "/articles/1"
+  def parent_path
```
Comment (brianjlandau) [inline app/models/comment.rb]: This is a view concern and should be a helper.

File: app/models/article.rb [hunk 17]
```diff
@@ -1,25 +1,51 @@
...
+  # def fill_user
+  #   curr_user = User.find(session[:user_id])
+  #   self.user_id = curr_user.id
+  # end
   
   def self.search(search, page)
   	paginate :per_page => 20, :page => page, :conditions => ['title || link like ?', "%#{search}%"]
 	end
+	
+	# This method is in comments as well, would like to condense this
```
Comment (brianjlandau) [inline app/models/article.rb]: Yes, you should create a `Commentable` module.

File: app/models/user.rb [hunk 18]
```diff
@@ -1,37 +1,27 @@
...
   has_secure_password
-  attr_accessible :username, :password, :password_confirmation, :role
+  attr_accessible :username, :password, :password_confirmation, :role, :about, :email
   
   # Relationships
   has_many :articles, :dependent => :destroy
 	
 	# Validations
-	validates_presence_of :username, :password, :password_confirmation, :role
+	validates_presence_of :username, :password, :password_confirmation, :role, :on => :create
```
Comment (brianjlandau) [inline app/models/user.rb]: Why only on create?

File: app/views/articles/index.html.erb [hunk 19]
```diff
@@ -1,22 +1,23 @@
...
-	</table>
-	<%= will_paginate @articles %>
-	
-	<%= form_tag articles_path, :method => 'get' do %>
-		<p>
-			<%= text_field_tag :search, params[:search] %>
-			<%= submit_tag "Search", :name => nil %>
-		</p>
-	<% end %>
+<% session[:return_to] = nil %>
```
Comment (brianjlandau) [inline app/views/articles/index.html.erb]: This shouldn't be part of the view. It should be in a controller. Also, why are you doing it here anyway?

File: app/views/articles/show.html.erb [hunk 20]
```diff
@@ -1,29 +1,9 @@
...
-</p>
-
 <p>
-  <b>Date added:</b>
-  <%= @article.date %>
+  	<h4><%= link_to @article.title, @article.link %> <%= "(" + @article.short_link + ")" %></h4>
+	<%= @article.points.to_s + " points by " %> <%= link_to @article.username, user_path(@article.user) %> <%="posted " + @article.created_at.to_s(:mdy) %>
 </p>
+	
+<%= render "partials/comments_form"%>
```
Comment (brianjlandau) [inline app/views/articles/show.html.erb]: `partials` isn't very descriptive. Also since this is related to comments it can probably live in the `comments` directory. So you could have a file named `app/views/comments/_form.html.erb` and `app/views/comments/_list.html.erb`.

When we do have common partials that aren't related to a specific resource, I think most of us have standardized on using a `app/views/shared` directory. I feel `shared` is a little more descriptive then `partials`.

File: app/controllers/application_controller.rb [hunk 21]
```diff
@@ -4,17 +4,31 @@ class ApplicationController < ActionController::Base
...
+    unless current_user
+      store_location
+      redirect_to login_path, alert: "You need to log in to view this page."
+      return false
+    end
+  end
+  
+  def store_location
```
Comment (brianjlandau) [inline app/controllers/application_controller.rb]: This should be a `before_filter` that is skipped when it needs to be skipped.

File: app/controllers/application_controller.rb [hunk 22]
```diff
@@ -4,17 +4,31 @@ class ApplicationController < ActionController::Base
...
   	current_user
   end
-  helper_method :logged_in?
-
-  def check_login
-  	redirect_to login_url, alert: "You need to log in to view this page." if current_user.nil?
+  
+  def require_user
+    unless current_user
+      store_location
```
Comment (brianjlandau) [inline app/controllers/application_controller.rb]: This shouldn't be called here, it should be it's own before filter.

File: app/views/layouts/application.html.erb [hunk 23]
```diff
@@ -32,7 +30,9 @@
 			<%= msg %>
 		</div>
     <% end %>
-
-	<%= yield %>
+	<div class="well">	
+		<%= session[:return_to] %>
```
Comment (brianjlandau) [inline app/views/layouts/application.html.erb]: This looks like a debugging line. Better to just use the logger for this kind of debugging.

File: app/views/partials/_comments_form.html.erb [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+		<fieldset>
+			<%= f.input :content %><br /><br />
+
+
+	      <%= f.submit "Add Comment", :class => 'btn btn-primary' %>
+
+	  </fieldset>
+	<% end %>
+<% else %>
+	<% session[:return_to] = request.url %>
```
Comment (brianjlandau) [inline app/views/partials/_comments_form.html.erb]: Using a before filter should take care of this.

File: app/views/partials/_comments_list.html.erb [hunk 25]
```diff
@@ -0,0 +1,16 @@
+<div id="comments">
+<% for comment in object %>
+	<div class="well">
+		<div class="comment">
+			<strong><%= link_to comment.username, user_path(comment.user) %></strong>
+			<em>on <%= comment.created_at.to_s(:mdyt) %> | <%= link_to "link", comment_path(comment) %></em><br />
+			<%= comment.content %><br /><br />
```
Comment (brianjlandau) [inline app/views/partials/_comments_list.html.erb]: Instead of using `<br />`'s you should think about using CSS to get the spacing you're looking for.

File: app/views/partials/_comments_list.html.erb [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+			<%= link_to "reply", comment_path(comment) %><br /><br />
+			<% if comment.comments.length != 0 %>
```
Comment (brianjlandau) [inline app/views/partials/_comments_list.html.erb]: `comment.comments.blank?` should work here.

File: app/views/users/_edit_form.html.erb [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+      <% @user.errors.full_messages.each do |msg| %>
+        <li><%= msg %></li>
+      <% end %>
+      </ul>
+    </div>
+  <% end %>
+	<p>
+    <%= f.input :about %><br />
+	<%= f.input :email %><br />
+	<% if request.url.include?("edit") %>
```
Comment (brianjlandau) [inline app/views/users/_edit_form.html.erb]: I wouldn't rely on the URL. Instead we should be able to introspect the user object and determine this:

``` ruby
if f.object.persisted?
```

Comment (brianjlandau) [inline app/views/users/_edit_form.html.erb]: Nevermind, just looked below and noticed that both paths can have a new and existing record. So why the 2 logic paths here? What am I missing?

File: config/environment.rb [hunk 28]
```diff
@@ -3,3 +3,7 @@
 
 # Initialize the rails application
 HaxxorNews::Application.initialize!
+
+
+Time::DATE_FORMATS[:mdy] = "%b %d, %Y"
```
Comment (brianjlandau) [inline config/environment.rb]: I would move this to an initializer file.

File: db/migrate/20120613145607_create_articles.rb [hunk 29]
```diff
@@ -3,7 +3,6 @@ def change
     create_table :articles do |t|
       t.string :title
       t.string :link
-      t.datetime :date, :default => Time.now
```
Comment (brianjlandau) [inline db/migrate/20120613145607_create_articles.rb:6]: This is ok, because you're the only one working on this. However, this is generally bad form, because other devs, FEDs may already have run this migration. In the future you should use a new migration to remove columns.

File: db/migrate/20120613201704_create_comments.rb [hunk 30]
```diff
@@ -5,7 +5,6 @@ def change
       t.integer :commentable_id
       t.string :commentable_type
       t.integer :user_id
-      t.datetime :date
```
Comment (brianjlandau) [inline db/migrate/20120613201704_create_comments.rb:8]: Same thing here about using a new independent migration to remove columns.

File: app/controllers/comments_controller.rb (hunk 10 above)
Comment (efatsi) [inline app/controllers/comments_controller.rb]: How do you feel about this?

``` ruby
redirect_to :back
```

It's more vague but definitely points to the right url.

File: app/models/article.rb (hunk 11 above)
Comment (efatsi) [inline app/models/article.rb]: this idea was letting single word strings through, might stick with the lengthy regex unless you strongly suggest otherwise

File: app/models/article.rb (hunk 15 above)
Comment (efatsi) [inline app/models/article.rb]: great idea, slightly adjusted to keep them small:

``` ruby
link.gsub(/\Ahttps?:\/\/(www.)?/, '').gsub(/\/.*/, '')
```

File: app/models/article.rb (hunk 17 above)
Comment (efatsi) [inline app/models/article.rb]: axed the method instead.

File: app/models/user.rb (hunk 18 above)
Comment (efatsi) [inline app/models/user.rb]: I want a user to be able to update their 'email' or 'about' section quickly and without needing password and confirmation. I'm thinking about redoing the auth in the near future though, I took a quick look at simplest_auth.

File: app/views/articles/index.html.erb (hunk 19 above)
Comment (efatsi) [inline app/views/articles/index.html.erb]: `:return_to` was lingering around in certain paths of use and thus directing the user to unexpected places. I figured this was a good place to reset it.

File: app/controllers/application_controller.rb (hunk 22 above)
Comment (efatsi) [inline app/controllers/application_controller.rb]: Are you suggesting the following?

``` ruby
before_filter :store_location, :only => [:require_user]
```

If so I don't see how that is better.

File: app/views/partials/_comments_form.html.erb (hunk 24 above)
Comment (efatsi) [inline app/views/partials/_comments_form.html.erb]: So should every `session[:return_to]` assignment be in a `before_filter`?

File: app/views/partials/_comments_list.html.erb (hunk 26 above)
Comment (efatsi) [inline app/views/partials/_comments_list.html.erb]: Nope. `.empty?` doesn't work either.

File: app/views/users/_edit_form.html.erb (hunk 27 above)
Comment (efatsi) [inline app/views/users/_edit_form.html.erb]: This was split so there would be a cancel button if you were editing info on the edit page, but given that you can edit your info in the show page there's no need for an edit page anyway. Problem solved.
Later I want to open up password changing but I'll cross that bridge when I get to it.

File: db/migrate/20120613145607_create_articles.rb (hunk 29 above)
Comment (efatsi) [inline db/migrate/20120613145607_create_articles.rb:6]: Okay, I just did this a bunch on a newer branch, I'll use migrations from now on though, it always seemed simpler/cleaner to change the migrations themselves but that makes sense.

File: app/controllers/comments_controller.rb (hunk 10 above)
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: If that works in every instance, that's exactly what you want and I would feel very good about it :smiley:. I figured you weren't doing that because in some cases the referer might not be where you want to send them to.

File: app/models/user.rb (hunk 18 above)
Comment (brianjlandau) [inline app/models/user.rb]: Ahhh, I see what you're doing. Well you always want to validate the presence of a `username`, because you don't want them to be able to remove it. Because the password fields are virtual attributes you could require them both on create, and then only require password confirmation if they have also have a password field, because that indicates they are actually trying to change their password. You should look at how, simplest_aut, authlogic, and devise handle this for inspiration.

To describe the requirement here: you always want a username on the model, and you want the password fields present when they are creating the record or when they are updating the password.

File: app/views/articles/index.html.erb (hunk 19 above)
Comment (brianjlandau) [inline app/views/articles/index.html.erb]: You should find solutions to this problem with good controller code, not view code.

File: app/controllers/application_controller.rb (hunk 22 above)
Comment (brianjlandau) [inline app/controllers/application_controller.rb]: No, you should just have:

``` ruby
before_filter :store_location
```

and then have:

``` ruby
skip_before_filter :store_location
```

in places you don't actually want to set that value.

File: app/views/partials/_comments_form.html.erb (hunk 24 above)
Comment (brianjlandau) [inline app/views/partials/_comments_form.html.erb]: Yes

Comment (brianjlandau) [inline app/views/partials/_comments_form.html.erb]: We can talk more about this tomorrow.

File: app/views/partials/_comments_list.html.erb (hunk 26 above)
Comment (brianjlandau) [inline app/views/partials/_comments_list.html.erb]: That's really odd. We should look into this more tomorrow. I'm interested to debug why those don't work.

File: db/migrate/20120613145607_create_articles.rb (hunk 29 above)
Comment (brianjlandau) [inline db/migrate/20120613145607_create_articles.rb:6]: It is when you're working alone. So for instance I might do this on a feature branch that I'm developing that I know no one is also working on. But for this project it might be good to think of this as something you might be sharing with others. That will give you some better real world practice.

File: app/controllers/application_controller.rb [hunk 31]
```diff
@@ -1,24 +1,18 @@
+require 'my_auth'
+require 'my_redirect'
```
Comment (brianjlandau) [inline app/controllers/application_controller.rb]: if you add this to `config/application.rb` you shouldn't need these require statements:

``` ruby
  config.autoload_paths += %W(#{config.root}/lib)
```

File: app/controllers/articles_controller.rb [hunk 32]
```diff
@@ -1,11 +1,13 @@
...
+  before_filter :assign_article, :only => [:show, :edit, :update, :destroy]
+  before_filter :require_user, :except => [:index, :show]
 
-  # load_and_authorize_resource
+  load_and_authorize_resource
 
   def index
-  	@articles = Article.search(params[:search], params[:page])		
+  	@articles = Article.search(params[:search], params[:page])
+  	session[:return_to] = nil
```
Comment (brianjlandau) [inline app/controllers/articles_controller.rb]: Why do you need to do this here? You shouldn't have to do this.

File: app/controllers/articles_controller.rb [hunk 33]
```diff
@@ -13,37 +15,30 @@ def index
     end
   end
 
-  # GET /articles/1
-  # GET /articles/1.json
   def show
-    @article = Article.find(params[:id])
-
+    session[:return_to] = request.url unless logged_in?
```
Comment (brianjlandau) [inline app/controllers/articles_controller.rb]: This should be taken care of by making `store_location` a before filter.

File: app/controllers/comments_controller.rb [hunk 34]
```diff
@@ -1,84 +1,58 @@
...
-  # GET /comments/1.json
+  
   def show
-    @comment = Comment.find(params[:id])
-
-    respond_to do |format|
-      format.html # show.html.erb
-      format.json { render json: @comment }
-    end
+    session[:return_to] = request.url unless logged_in?
```
Comment (brianjlandau) [inline app/controllers/comments_controller.rb]: This should be taken care of by making store_location a before filter.

File: app/views/layouts/application.html.erb [hunk 35]
```diff
@@ -32,7 +30,9 @@
 			<%= msg %>
 		</div>
     <% end %>
-
-	<%= yield %>
+	<div class="well">	
+		<% logger.debug "#{session[:return_to]}" %>
```
Comment (brianjlandau) [inline app/views/layouts/application.html.erb]: `logger.debug session[:return_to].to_s`

File: config/environment.rb [hunk 36]
```diff
@@ -3,3 +3,9 @@
...
+Time::DATE_FORMATS[:mdyt] = "%b %d, %Y at %H:%M"
+
+Rails.logger = Logger.new(STDOUT)
```
Comment (brianjlandau) [inline config/environment.rb]: Why did you change to `STDOUT`? Also this should be in an environment file, like `development.rb`

File: lib/my_auth.rb [hunk 37]
```diff
@@ -0,0 +1,10 @@
+module MyAuth
```
Comment (brianjlandau) [inline lib/my_auth.rb]: I would rename this to just `Authentication`. You could always namespace it under your application module name (`HaxxorNews` or whatever).

I would also take the contents of `MyRedirect` and put it in here. They are all related to authentication.

**Comment by efatsi** [comment]
Thanks for the help Brian, merging it in now.
